pkgs/os-specific/linux/kernel/linux-testing-bcachefs.nix

```diff
@@ -1,18 +1,29 @@
-{ lib, buildPackages, fetchFromGitHub, fetchpatch, perl, buildLinux, ... } @ args:
+{ lib
+, fetchpatch
+, kernel
+, date ? "2021-05-20"
+, commit ? "ac91176cb1f48e218d19026977ebe52078297c32"
+, diffHash ? "1flkvq7v35ay6x9z9519bylah8l1a787s95gq6bn30izwzq1vx6p"
+, kernelPatches ? [] # will always be supplied by callPackage because it is also a top-level attribute
+, argsOverride ? {}
+, ...
+} @ args:
 
-buildLinux (args // {
-  version = "5.9.0-2020.11.20";
-  modDirVersion = "5.9.0";
+kernel.override ( args // {
 
-  src = fetchFromGitHub {
-    owner = "koverstreet";
-    repo = "bcachefs";
-    # commit does not exist on any branch on the target repository
-    rev = "6a505b63ed3003faf5000f19fd08bbd477d93fbc";
-    sha256 = "1rf34gzv9npafp1c3i6lymk3b0gnqp4rb0wl33pw6yrpgnsry3cc";
-  };
+  argsOverride = {
+    version = "${kernel.version}-${date}";
+  } // argsOverride;
 
-  extraConfig = "BCACHEFS_FS m";
+  kernelPatches = let bcachefs-patch = {
+      name = "bcachefs-${commit}";
+      patch = fetchpatch {
+        name = "bcachefs-${commit}.diff";
+        url = "https://evilpiepirate.org/git/bcachefs.git/rawdiff/?id=${commit}&id2=v${lib.versions.major kernel.version}.${lib.versions.minor kernel.version}";
+        sha256 = diffHash;
+      };
+      extraConfig = "BCACHEFS_FS m";
+    }; in [ bcachefs-patch ] ++ lib.remove bcachefs-patch kernelPatches;
```

Comment:
If someone could explain to me why the patch is applied more than once without this line I'd be glad. ^^

Reply:
does it help to remove the `rec` below, and use `linux_5_11.kernelPatches` instead of `kernel.kernelPatches` ?

Reply:
No, that is unrelated.

Reply:
Are you sure the patch is applied twice otherwise / does it fail or something?

Reply:
The patch is definitely applied more than once without removing it from the passed kernelPatches. I get that you're confirming because it does not make sense to me either. ^^

Reply:
(its normal that the patch appears twice in the build log I think, because `linux-config` is built first, where the patch is also applied)

Reply:
It is definitely applied more than once to the same source because the second attempt to do so fails.

Reply:
No, I meant that I generally see patches being applied twice, once for the derivation generating the config, and once for the build proper. I actually don't see the patch in question in the build log at all..

Reply:
ah never mind, that was my mistake.. I can see it now, but it is only applied once for me (and succeeds) I think?

Reply:
Well, because of the filtering. You should find that without filtering the build fails.

Reply:
Nope, it runs fine for me, even without the filter

Reply:
Turns out you're right, and I absolutely cannot tell you what else originally caused the double application. For me that replaces one mystery with another, but for you it's solved. 😅

Reply:
Apparently this issue does NOT occur when building the kernel as a package, but DOES occur when using it in NixOS. Will need to investigate further.

Reply:
I still can't really find out why this is necessary but I will leave it in for now given that there's precedent in the rt kernel (where I originally got it from). Maybe @orivej, who first introduced the pattern, can chime in and explain.

Reply:
The issue appears to be, that the code in the `boot.kernelPackages` option assumes that the `kernelPatches` attribute of a kernel coincides with the one that was passed into the function generating it. In cases where we modify `kernelPatches` like it happens here, this does not hold.

Reply:
ah thats probably hard to do without hard coding the commit hash etc everywhere.
I will try to think of a proper solution for the general problem, but as far as this pr is concerned the current approach (removing it after the fact) seems fine.

Reply:
So this is tricky to do, but I have a commit here https://github.com/xaverdh/nixpkgs/commit/85e2a3809aff54970065d22d4aa23f57b1d4f03a on top of the changes in https://github.com/NixOS/nixpkgs/pull/123401 that should hopefully work (didn't test yet).

Reply:
Maybe its not that tricky after all. It appears that `override` can also take a function (passing the original arguments). Not sure why its not documented. Can you try if https://github.com/NixOS/nixpkgs/pull/124486 does the trick?

Reply:
That works indeed. Thanks for fixing this up!